servers/src/main/java/tachyon/master/TachyonMaster.java

```diff
@@ -203,168 +188,168 @@ public String getWebBindHost() {
+    return mRawTableMaster;
   }
 
   /**
-   * Get whether the system is in zookeeper mode, for unit test only.
+   * Get internal {@link BlockMaster}, for unit test only.
+   */
+  public BlockMaster getBlockMaster() {
+    return mBlockMaster;
+  }
+
+  /**
+   * Get the millisecond when Tachyon Master starts serving, return -1 when not started.
+   */
+  public long getStarttimeMs() {
+    return mStartTimeMs;
+  }
+
+  /**
+   * Get whether the system is serving the rpc server.
    *
-   * @return true if the master is under zookeeper mode, false otherwise.
+   * @return true if the system is the leader (serving the rpc server), false otherwise.
    */
-  boolean isZookeeperMode() {
-    return mZookeeperMode;
+  boolean isServing() {
+    return mIsServing;
   }
 
-  private void connectToUFS() throws IOException {
-    String ufsAddress =
-        mTachyonConf.get(Constants.UNDERFS_ADDRESS);
-    UnderFileSystem ufs = UnderFileSystem.get(ufsAddress, mTachyonConf);
-    ufs.connectFromMaster(mTachyonConf,
-        NetworkAddressUtils.getConnectHost(ServiceType.MASTER_RPC, mTachyonConf));
+  /**
+   * Start the Tachyon master server.
+   */
+  public void start() throws Exception {
+    startMasters(true);
+    startServing();
   }
 
-  private void setup() throws IOException, TTransportException {
-    connectToUFS();
-    if (mZookeeperMode) {
-      mEditLogProcessor.stop();
+  /**
+   * Stop the Tachyon master server. Should only be called by tests.
+   */
+  public void stop() throws Exception {
+    if (mIsServing) {
+      LOG.info("Stopping Tachyon Master @ " + mMasterAddress);
+      stopServing();
+      stopMasters();
+      mTServerSocket.close();
+      mIsServing = false;
     }
-    mMasterInfo.init();
+  }
 
-    mWebServer =
-        new MasterUIWebServer(ServiceType.MASTER_WEB, NetworkAddressUtils.getBindAddress(
-            ServiceType.MASTER_WEB, mTachyonConf), mMasterInfo, mTachyonConf);
+  protected void startMasters(boolean isLeader) {
+    try {
+      connectToUFS();
 
-    mMasterServiceHandler = new MasterServiceHandler(mMasterInfo);
-    MasterService.Processor<MasterServiceHandler> masterServiceProcessor =
-        new MasterService.Processor<MasterServiceHandler>(mMasterServiceHandler);
+      mBlockMaster.start(isLeader);
+      mFileSystemMaster.start(isLeader);
+      mRawTableMaster.start(isLeader);
 
-    mMasterServiceServer =
-        new TThreadPoolServer(new TThreadPoolServer.Args(mServerTServerSocket)
-            .maxWorkerThreads(mMaxWorkerThreads).minWorkerThreads(mMinWorkerThreads)
-            .processor(masterServiceProcessor).transportFactory(new TFramedTransport.Factory())
-            .protocolFactory(new TBinaryProtocol.Factory(true, true)));
+    } catch (IOException e) {
+      LOG.error(e.getMessage(), e);
+      throw Throwables.propagate(e);
+    }
+  }
 
-    mIsStarted = true;
+  protected void stopMasters() {
+    try {
+      mBlockMaster.stop();
```

Comment:
why `startMasters` connect to ufs but `stopMasters` doesn't cleanup ufs connection?